### PR TITLE
[New Feature][Solace Integration-03] Add Solace Specific notifier and flow implementation.

### PR DESCRIPTION
**Pull request by Chamindu36**
### Purpose
This PR will complete the implementation of the following tasks.

_1. Design for externalizing the Solace broker deployments and creations.
2. Support deploying created AsyncAPI into Solace environment through Notifier Approach.
3. Add Deployer interface and implementation to support External brokers.
4. Add Solace deployer implementation to support deploy and undeploy tasks of API
5. Solace API deletion and Life cycle change flow implementation with APINotifier
6. Support creating an application using the deployed API in DevPortal and Solace Broker through subscription notifier. 
7. Add solace key gen notifier to support key generation.
8. Support unsubscribing to Solace APIs via the Dev portal.
9. Support deleting an application in the Solace Broker.
10. Activate newly added notifiers and Deployers in server startup_

### Goal
Fixes : 
https://github.com/wso2/product-apim/issues/11650
https://github.com/wso2/product-apim/issues/11652
https://github.com/wso2/product-apim/issues/11799
https://github.com/wso2/product-apim/issues/11717
https://github.com/wso2/product-apim/issues/11760
https://github.com/wso2/product-apim/issues/11718
https://github.com/wso2/product-apim/issues/11719
https://github.com/wso2/product-apim/issues/11761

### Related PRs 
Marge after the following PR - 
https://github.com/wso2/carbon-apimgt/pull/10840

### CODE REVIEW

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIManagerConfiguration.java [hunk 1]
```diff
@@ -626,6 +523,135 @@ private void readChildElements(OMElement serverConfig,
...
+            environment.setShowInConsole(true);
+        }
+        String isDefault = environmentElem.getAttributeValue(new QName("isDefault"));
+        if (isDefault != null) {
+            environment.setDefault(Boolean.parseBoolean(isDefault));
+        } else {
+            environment.setDefault(false);
+        }
+        environment.setName(APIUtil.replaceSystemProperty(
+                environmentElem.getFirstChildWithName(new QName(
```
Comment (dnwick) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIManagerConfiguration.java:548]: Didnt we fixed these formatting issues ?

**Review by dnwick — APPROVED** [review]

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/notifier/ExternallyDeployedApiNotifier.java [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+     * @throws NotifierException if error occurs when undeploying APIs from external gateway
+     */
+    private void undeployApiWhenRetiring(APIEvent apiEvent) throws NotifierException {
+
+        apiMgtDAO = ApiMgtDAO.getInstance();
+        Map<String, Environment> gatewayEnvironments = APIUtil.getReadOnlyGatewayEnvironments();
+        boolean deletedFromSolace;
+        String apiId = apiEvent.getUuid();
+
+        if (!apiEvent.getApiStatus().equals(APIConstants.RETIRED)) {
```
Comment (tharindu1st) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/notifier/ExternallyDeployedApiNotifier.java:87]: other way round

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/notifier/ExternallyDeployedApiNotifier.java [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+            List<APIRevisionDeployment> test = apiMgtDAO.getAPIRevisionDeploymentsByApiUUID(apiId);
+
+            for (APIRevisionDeployment deployment : test) {
+                String deploymentEnv = deployment.getDeployment();
+                if (gatewayEnvironments.containsKey(deploymentEnv)) {
+                    ExternalGatewayDeployer deployer = ServiceReferenceHolder.getInstance().getExternalGatewayDeployer
+                            (gatewayEnvironments.get(deploymentEnv).getProvider());
+                    if (deployer != null) {
+                        try {
+                            deletedFromSolace = deployer.undeployWhenRetire(api, gatewayEnvironments.get(deploymentEnv));
```
Comment (tharindu1st) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/notifier/ExternallyDeployedApiNotifier.java:104]: deletedFromSolace -> deleted

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/notifier/ExternallyDeployedApiNotifier.java [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+        try {
+            List<APIRevisionDeployment> test = apiMgtDAO.getAPIRevisionDeploymentsByApiUUID(apiId);
+            for (APIRevisionDeployment deployment : test) {
+                String deploymentEnv = deployment.getDeployment();
+                if (gatewayEnvironments.containsKey(deploymentEnv)) {
+                    ExternalGatewayDeployer deployer = ServiceReferenceHolder.getInstance().getExternalGatewayDeployer
+                            (gatewayEnvironments.get(deploymentEnv).getProvider());
+                    if (deployer != null) {
+                        try {
+                            deletedFromSolace = deployer.undeploy(apiEvent.getApiName(), apiEvent.getApiVersion(),
```
Comment (tharindu1st) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/notifier/ExternallyDeployedApiNotifier.java:141]: deletedFromSolace -> deleted

File: components/apimgt/org.wso2.carbon.apimgt.solace/src/main/java/org/wso2/carbon/apimgt/solace/SolaceAdminApis.java [hunk 5]
```diff
@@ -0,0 +1,824 @@
...
+     * @param apiNameForRegistration name of the Solace API product
+     * @param aai20Document          Async definition of the Solace API
+     * @return CloseableHttpResponse of the POST call
+     */
+    public CloseableHttpResponse createAPIProduct(String organization, String environment, Aai20Document aai20Document,
+                                         String apiProductName, String apiNameForRegistration) {
+        URL serviceEndpointURL = new URL(baseUrl);
+        HttpClient httpClient = APIUtil.getHttpClient(serviceEndpointURL.getPort(), serviceEndpointURL.getProtocol());
+        HttpPost httpPost = new HttpPost(baseUrl + "/" + organization + "/apiProducts");
+        httpPost.setHeader(HttpHeaders.AUTHORIZATION, "Basic " + getEncoding());
```
Comment (tharindu1st) [inline components/apimgt/org.wso2.carbon.apimgt.solace/src/main/java/org/wso2/carbon/apimgt/solace/SolaceAdminApis.java:159]: is this correct getEncoding()

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/deployer/ExternalGatewayDeployerImpl.java [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+package org.wso2.carbon.apimgt.impl.deployer;
+
+import org.wso2.carbon.apimgt.api.model.API;
+import org.wso2.carbon.apimgt.api.model.Environment;
+import org.wso2.carbon.apimgt.impl.deployer.exceptions.DeployerException;
+
+public class ExternalGatewayDeployerImpl implements ExternalGatewayDeployer {
```
[the remark this answers is not in the document]
Comment (Chamindu36) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/deployer/ExternalGatewayDeployerImpl.java:22]: Fixed with https://github.com/wso2/carbon-apimgt/pull/10841/commits/9ecce20c70204a6168d394dd441de6b444223728

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/deployer/exceptions/DeployerException.java [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+package org.wso2.carbon.apimgt.impl.deployer.exceptions;
+
+public class DeployerException extends Exception {
```
[the remark this answers is not in the document]
Comment (Chamindu36) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/deployer/exceptions/DeployerException.java:21]: Fixed with https://github.com/wso2/carbon-apimgt/pull/10841/commits/9ecce20c70204a6168d394dd441de6b444223728

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/notifier/ExternalGatewayNotifier.java [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+     * Process gateway notifier events related to External gateway deployments
+     *
+     * @param event related to deployments
+     * @throws NotifierException if error occurs when casting event
+     */
+    private void process (Event event) throws NotifierException {
+        DeployAPIInGatewayEvent deployAPIInGatewayEvent;
+        try {
+            deployAPIInGatewayEvent = (DeployAPIInGatewayEvent) event;
+        } catch (ExceptionInInitializerError e) {
```
[the remark this answers is not in the document]
Comment (Chamindu36) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/notifier/ExternalGatewayNotifier.java]: Fixed with https://github.com/wso2/carbon-apimgt/pull/10841/commits/9ecce20c70204a6168d394dd441de6b444223728

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/notifier/ExternalGatewayNotifier.java [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+                    getThreadLocalCarbonContext().getUsername());
+            API api = apiProvider.getAPIbyUUID(apiId, apiMgtDAO.getOrganizationByAPIUUID(apiId));
+
+            for (String deploymentEnv : gateways) {
+                if (gatewayEnvironments.containsKey(deploymentEnv)) {
+                    ExternalGatewayDeployer deployer = ServiceReferenceHolder.getInstance().getExternalGatewayDeployer
+                            (gatewayEnvironments.get(deploymentEnv).getProvider());
+                    if ( deployer!= null) {
+                        try {
+                            deployedToSolace = deployer.deploy(api, gatewayEnvironments.get(deploymentEnv));
```
[the remark this answers is not in the document]
Comment (Chamindu36) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/notifier/ExternalGatewayNotifier.java]: Fixed with https://github.com/wso2/carbon-apimgt/pull/10841/commits/9ecce20c70204a6168d394dd441de6b444223728

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/notifier/ExternalGatewayNotifier.java [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+                    getThreadLocalCarbonContext().getUsername());
+            API api = apiProvider.getAPIbyUUID(apiId, apiMgtDAO.getOrganizationByAPIUUID(apiId));
+
+            for (String deploymentEnv : gateways) {
+                if (gatewayEnvironments.containsKey(deploymentEnv)) {
+                    ExternalGatewayDeployer deployer = ServiceReferenceHolder.getInstance().getExternalGatewayDeployer
+                            (gatewayEnvironments.get(deploymentEnv).getProvider());
+                    if (deployer != null) {
+                        try {
+                            deletedFromSolace = deployer.undeploy(api.getId().getName(), api.getId().getVersion(),
```
[the remark this answers is not in the document]
Comment (Chamindu36) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/notifier/ExternalGatewayNotifier.java]: Fixed with https://github.com/wso2/carbon-apimgt/pull/10841/commits/9ecce20c70204a6168d394dd441de6b444223728

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/notifier/ExternallyDeployedApiNotifier.java (hunk 2 above)
Comment (Chamindu36) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/notifier/ExternallyDeployedApiNotifier.java]: Fixed with https://github.com/wso2/carbon-apimgt/pull/10841/commits/9ecce20c70204a6168d394dd441de6b444223728

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/notifier/ExternallyDeployedApiNotifier.java (hunk 3 above)
Comment (Chamindu36) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/notifier/ExternallyDeployedApiNotifier.java]: Fixed with https://github.com/wso2/carbon-apimgt/pull/10841/commits/9ecce20c70204a6168d394dd441de6b444223728

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/notifier/ExternallyDeployedApiNotifier.java (hunk 4 above)
Comment (Chamindu36) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/notifier/ExternallyDeployedApiNotifier.java]: Fixed with https://github.com/wso2/carbon-apimgt/pull/10841/commits/9ecce20c70204a6168d394dd441de6b444223728

File: components/apimgt/org.wso2.carbon.apimgt.solace/src/main/java/org/wso2/carbon/apimgt/solace/SolaceAdminApis.java (hunk 5 above)
Comment (Chamindu36) [reply components/apimgt/org.wso2.carbon.apimgt.solace/src/main/java/org/wso2/carbon/apimgt/solace/SolaceAdminApis.java:159]: @tharindu1st Yes. This is a newly written private method that is used for avoiding repetition in HTTP request creation. Not the base64.encoding method.

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/deployer/ExternalGatewayDeployer.java [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+ *  limitations under the License.
+ */
+
+package org.wso2.carbon.apimgt.impl.deployer;
+
+import org.wso2.carbon.apimgt.api.model.API;
+import org.wso2.carbon.apimgt.api.model.Environment;
+import org.wso2.carbon.apimgt.impl.deployer.exceptions.DeployerException;
+
+public interface ExternalGatewayDeployer {
```
[the remark this answers is not in the document]
Comment (Chamindu36) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/deployer/ExternalGatewayDeployer.java:26]: Fixed with https://github.com/wso2/carbon-apimgt/pull/10841/commits/9ecce20c70204a6168d394dd441de6b444223728

**Review by tharindu1st — APPROVED** [review]

**Pull request merged**